graphql/src/execution/ast.rs

```diff
@@ -0,0 +1,349 @@
+use std::{collections::HashSet, ops::Deref};
+
+use graph::{
+    components::store::EntityType,
+    data::graphql::{DocumentExt, ObjectOrInterface},
+    prelude::{anyhow, q, r, s, QueryExecutionError, Schema, ValueMap},
+};
+use graphql_parser::Pos;
+
+use crate::schema::ast::ObjectType;
+
+/// A selection set is a table that maps object types to the fields that
+/// should be selected for objects of that type. The types are always
+/// concrete object types, never interface or union types. When a
+/// `SelectionSet` is constructed, fragments must already have been resolved
+/// as it only allows using fields.
+///
+/// The set of types that a `SelectionSet` can accommodate must be set at
+/// the time the `SelectionSet` is constructed. It is not possible to add
+/// more types to it, but it is possible to add fields for all known types
+/// or only some of them
+#[derive(Debug, Clone, PartialEq)]
+pub struct SelectionSet {
+    // Map object types to the list of fields that should be selected for
+    // them
+    items: Vec<(String, Vec<Field>)>,
```

Comment:
Is `items.len() > 1` only when the selection is over a union or interface? If so it would be clarifying to note that in the comment.

Reply:
It would be nice to use a `type` for strings that represent object names (or type names in general, whatever makes more sense).

Reply:
I've wanted to use something like `&s::ObjectType` here, but that's not possible for various reasons. I've now resolved to do the next best thing, which is keeping a copy of all object types in `ApiSchema`, and wrapping them in an `Arc`. With that, a `SelectionSet` now has basically `items: Vec<(Arc<s::ObjectType>, Vec<Field>)>` (in reality a little more complicated to work around other shortcomings of `s::ObjectType`)

I am curios what you think of that approach.

Reply:
Looks good! I see that saved some calls to `get_object_type_definition`.

Reply:
Added a comment to clarify when `items.len() > 1`